llvm/lib/Analysis/Loads.cpp

```diff
@@ -856,16 +856,19 @@ bool llvm::canReplacePointersIfEqual(const Value *From, const Value *To,
   return isPointerAlwaysReplaceable(From, To, DL);
 }
 
-bool llvm::isDereferenceableReadOnlyLoop(
+bool llvm::isReadOnlyLoopWithSafeOrSpeculativeLoads(
     Loop *L, ScalarEvolution *SE, DominatorTree *DT, AssumptionCache *AC,
+    SmallVectorImpl<LoadInst *> *SpeculativeLoads,
```

Comment:
Nit, the term "speculative" reminds me of `isSafeToSpeculativelyExecute` which is used for div/rem etc. I think it would sound better if we just used the existing terminology, so called this function something like `isReadOnlyLoop` and renamed the argument to

```suggestion
    SmallVectorImpl<LoadInst *> *NonDereferenceableAndAlignedLoads,
```

Since I guess any load where `isDereferenceableAndAlignedInLoop` returns true is also technically speculative.